### PR TITLE
[python] Append-mode pre-check logic

### PR DESCRIPTION
**Pull request by johnkerl**
**Issue and/or context:**

Tracking issue: #1278 

**Changes:**

* Create the `Signature` class which will be used for append-mode pre-check on an upcoming PR
* Unit-test that

**Notes for Reviewer:**

* This is a split-out PR -- created for mercifulness to reviewers -- as a line-count reduction from raw material on #1498.
* Question for @aaronwolen @mojaveazure : If an input file has raw and non-raw data do we want them both to be ingested? Or just one or the other? This is actually a design question -- which I'll also put in our tracking doc.

### CODE REVIEW

File: apis/python/src/tiledbsoma/io/registration/signatures.py
Comment (thetorpedodog) [inline apis/python/src/tiledbsoma/io/registration/signatures.py:31]: Optional: This could be implemented with a little less repetition with a dictionary:

```python
_TYPE_MAP = {
  "large_string": "string",
  "large_binary": "binary",
}

def _stringify_type(t: pa.DataType) -> str:
  str_t = str(t)
  return _TYPE_MAP.get(str_t, str_t)
```

I would also recommend pulling the `stringify_type` function to the root level; since it doesn’t use anything from its closure it doesn’t really gain anything from being in one.

File: apis/python/src/tiledbsoma/io/registration/signatures.py
Comment (thetorpedodog) [inline apis/python/src/tiledbsoma/io/registration/signatures.py:108]: snake case `from_anndata` etc.?

File: apis/python/src/tiledbsoma/io/registration/signatures.py
Comment (thetorpedodog) [inline apis/python/src/tiledbsoma/io/registration/signatures.py:245]: This seems more like you would want to handle it as a `check_compatibility` function that maybe raised an exception of some kind. The signature as-is has the risk of being misunderstood; a person could write

```python
if compatible(some_sigs):
  do_dangerous_thing(...)
```

and since a 2-tuple is non-empty, it would always evaluate as truthy and `do_dangerous_thing`

File: apis/python/src/tiledbsoma/io/registration/signatures.py
Comment (thetorpedodog) [inline apis/python/src/tiledbsoma/io/registration/signatures.py:247]: any number of signatures

File: apis/python/src/tiledbsoma/io/registration/signatures.py
Comment (thetorpedodog) [inline apis/python/src/tiledbsoma/io/registration/signatures.py:288]: Can this be replaced with `self == other`?

File: apis/python/src/tiledbsoma/io/registration/signatures.py
Comment (thetorpedodog) [inline apis/python/src/tiledbsoma/io/registration/signatures.py:304]: You can save some typing with `cls(**dikt)`. Also maybe worth declaring the class as `kw_only=True`, because it seems like there’s no obviously correct order for these.

File: apis/python/src/tiledbsoma/_arrow_types.py
Comment (thetorpedodog) [inline apis/python/src/tiledbsoma/_arrow_types.py:171]: I see this is just moved code, so you don’t need to make this change if you don’t want to, but:

Dataframes implement the `dict`-like `.items` method, so this can be `for name, col in df.items()` to make some usage of `df[k]` a bit more readable

**Review by thetorpedodog — COMMENTED** [review]
Overall the structure of this code is good; I do have some suggestions and a few concerns about the API right now, though.

File: apis/python/src/tiledbsoma/_arrow_types.py
Comment (johnkerl) [reply apis/python/src/tiledbsoma/_arrow_types.py:171]: @thetorpedodog no b/c I want `df[k]` as left-hand sides to mutate

**Comment by johnkerl** [comment]
@thetorpedodog ready for round two

File: apis/python/src/tiledbsoma/io/registration/signatures.py
Comment (thetorpedodog) [inline apis/python/src/tiledbsoma/io/registration/signatures.py:27]: this can either be a module-level constant or make it a regular variable name like `equivalences`

File: apis/python/src/tiledbsoma/io/registration/signatures.py
Comment (thetorpedodog) [inline apis/python/src/tiledbsoma/io/registration/signatures.py:299]: nit: `to_json`/`from_json`

File: apis/python/src/tiledbsoma/io/registration/signatures.py
Comment (thetorpedodog) [inline apis/python/src/tiledbsoma/io/registration/signatures.py:302]: This seems like it could be `json.dumps(attrs.asdict(self), ...)` rather than setting `default=`. Am I missing something?

(If there is a reason to set `default`, then it can be shortened to `default=attrs.asdict`, since that’s all the lambda does.)

File: apis/python/src/tiledbsoma/io/registration/signatures.py
Comment (thetorpedodog) [inline apis/python/src/tiledbsoma/io/registration/signatures.py:77]: I would [use `kw_only` here](https://www.attrs.org/en/stable/api-attr.html#attr.s:~:text=kw_only%20,__init__) to avoid accidentally passing things in as positional arguments. 

```python
@attrs.define(kw_only=True)
class Signature:
  ...
```

**Review by thetorpedodog — APPROVED** [review]
Other than the few comments below, looks good to me style-wise. The logic seems reasonable but I’m not *super* familiar; you may wish to get another reviewer to give that a once-over (but not mandatory).

File: apis/python/src/tiledbsoma/io/registration/signatures.py
Comment (johnkerl) [reply apis/python/src/tiledbsoma/io/registration/signatures.py:302]: It is necessary to exist; I get errors without. I'll simplify the syntax as you suggest.

**Comment by johnkerl** [comment]
@aaronwolen @mojaveazure any final concerns before I merge? This is of course a multi-PR (at least 3-PR) WIP and I'm happy to take feedback before or after the merge ...

**Comment by johnkerl** [comment]
Merging -- happy to take belated feedback as a follow-on PR

**Pull request merged**
